### PR TITLE
Change the async interface to be an attribute of the blocking functions

### PR DESCRIPTION
**Pull request by freider**
This deprecates `.create_async()` "purely async" interfaces in favor of putting `.aio` attributes on emitted blocking functions/methods (if they are determined to be async).

Instead of
```
aio_func = create_async(_func)
await aio_func(some_arg)
```
You would now access the async interface from the blocking one like this:
```
func = create_blocking(_func)
await func.aio(some_arg)
```

This has the advantage that there will no longer be two distinct wrapper types used by users of the library, all wrapper types will be blocking, but be callable as async where applicable by using the .aio attribute on functions.

For the type stubs to work with this, all async-enabled functions/methods are now emitted as "private" typing.Protocols, with visible instances, having both the `def __call__` and `async def aio` methods, eg.:
```
class __func_spec(typing.Protocol):
    def __call__(self, some_arg): ...
    async def aio(self, some_arg): ...

func: __func_spec 
```

During a transition period, we keep the public `.create_async()` around, as well as a hidden new `Interface._ASYNC_WITH_BLOCKING_TYPES` interface, which represents the new async interface on methods (which now runs async but *returns* blocking entities). 

After the transition period, we'll remove the ASYNC interface, and have only one public wrapper interface which is both blocking and async as per above.

TODO:
* Translate/add tests to use the .aio method instead of the old async interface

### CODE REVIEW

File: test/synchronicity_test.py [hunk 1]
```diff
@@ -66,6 +66,32 @@ async def test_function_async():
     assert await coro == 1764
 
 
+@pytest.mark.asyncio
+async def test_function_async_as_function_attribute():
+    s = Synchronizer()
+    t0 = time.time()
+    f_s = s.create_blocking(f).aio
```
Comment (erikbern) [inline test/synchronicity_test.py:73]: instead of `create_blocking` maybe we should just call it `create`?

File: test/type_stub_test.py [hunk 2]
```diff
@@ -257,16 +258,21 @@ async def clone(foo: "_Foo") -> "_Foo":
...
+
+    assert "class __get_foo_spec(typing_extensions.Protocol):" in src
+    assert "    def __call__(self, foo: Foo) -> Foo" in src
+    assert "    async def aio(self, *args, **kwargs) -> Foo" in src
+    assert "get_foo: __get_foo_spec"
 
 
 def test_synchronicity_self_ref():
     src = _class_source(Foo)
+    print(src)
```
Comment (erikbern) [inline test/type_stub_test.py:275]: debug print

**Comment by erikbern** [comment]
Nice!

It looks like the tracebacks tests are failing. Those can be a bit hard to fix. I'm OK disabling those tests temporarily since this PR means a lot of code will be deprecated. It might be easier to revisit those tests once we've removed all the deprecated code later.

**Review by erikbern — APPROVED** [review]

File: test/synchronicity_test.py (hunk 1 above)
Comment (freider) [reply test/synchronicity_test.py:73]: We already have a deprecated `.create()` which returns:
```
{
            Interface.ASYNC: self.create_async(obj),
            Interface.BLOCKING: self.create_blocking(obj),
        }
```
in order to support old modal clients that didn't pin the synchronicity version. We'd have to wait until we phase out those modal versions until we reuse .create for something else right?

Comment (freider) [reply test/synchronicity_test.py:73]: We have had the `"Using latest synchronicity with an old interface - please upgrade to latest modal-client!"` warning there though, so perhaps it's acceptable to phase out now?

Comment (erikbern) [reply test/synchronicity_test.py:75]: Yeah seems fine to phase out!

Comment (erikbern) [reply test/synchronicity_test.py:75]: PR here: https://github.com/modal-labs/synchronicity/pull/100

**Pull request merged**